atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/RowColumnRangeExtractor.java

```diff
@@ -68,57 +73,53 @@ public Map<byte[], Integer> getRowsToRawColumnCount() {
+    static RowColumnRangeResult extract(
+            Collection<byte[]> canonicalRows,
+            Map<ByteBuffer, List<ColumnOrSuperColumn>> colsByKey,
+            long startTs,
+            MetricsManager metricsManager) {
+        IdentityHashMap<byte[], LinkedHashMap<Cell, Value>> collector = new IdentityHashMap<>(canonicalRows.size());
+        IdentityHashMap<byte[], Column> rowsToLastCompositeColumns = new IdentityHashMap<>(canonicalRows.size());
+        IdentityHashMap<byte[], Integer> rowsToRawColumnCount = new IdentityHashMap<>(canonicalRows.size());
+        Set<byte[]> emptyRows = Collections.newSetFromMap(new IdentityHashMap<>(0));
 
-    RowColumnRangeExtractor(MetricsManager metricsManager) {
-        notLatestVisibleValueCellFilterCounter = metricsManager.registerOrGetCounter(
-                RowColumnRangeExtractor.class, AtlasDbMetricNames.CellFilterMetrics.NOT_LATEST_VISIBLE_VALUE);
-    }
+        // lazily create counter to avoid overhead when not needed
+        Supplier<Counter> notLatestVisibleValueCellFilterCounter =
+                Suppliers.memoize(() -> metricsManager.registerOrGetCounter(
+                        RowColumnRangeExtractor.class, CellFilterMetrics.NOT_LATEST_VISIBLE_VALUE));
 
-    public void extractResults(
-            Iterable<byte[]> canonicalRows, Map<ByteBuffer, List<ColumnOrSuperColumn>> colsByKey, long startTs) {
         // Make sure returned maps are keyed by the given rows
         Map<ByteBuffer, byte[]> canonicalRowsByHash = Maps.uniqueIndex(canonicalRows, ByteBuffer::wrap);
         for (Map.Entry<ByteBuffer, List<ColumnOrSuperColumn>> colEntry : colsByKey.entrySet()) {
             byte[] rawRow = CassandraKeyValueServices.getBytesFromByteBuffer(colEntry.getKey());
             byte[] row = canonicalRowsByHash.get(ByteBuffer.wrap(rawRow));
             List<ColumnOrSuperColumn> columns = colEntry.getValue();
 
-            if (!columns.isEmpty()) {
+            if (columns.isEmpty()) {
+                emptyRows.add(row);
+            } else {
                 rowsToLastCompositeColumns.put(
                         row, columns.get(columns.size() - 1).getColumn());
-            } else {
-                emptyRows.add(row);
             }
             rowsToRawColumnCount.put(row, columns.size());
             for (ColumnOrSuperColumn c : columns) {
                 Pair<byte[], Long> pair = CassandraKeyValueServices.decomposeName(c.getColumn());
-                internalExtractResult(startTs, row, pair.lhSide, c.getColumn().getValue(), pair.rhSide);
+                long ts = pair.rhSide;
+                if (ts < startTs) {
+                    Cell cell = Cell.create(row, pair.lhSide);
+                    LinkedHashMap<Cell, Value> cellToValue = collector.get(row);
+                    if (cellToValue == null) {
+                        cellToValue = collector.computeIfAbsent(row, _b -> new LinkedHashMap<>(1));
+                    }
```

Comment:
(But this is correct too!)

Reply:
I went back and forth a bit on this, and ultimately went with `cellToValue.containsKey` & `cellToValue.put` rather than `putIfAbsent` to avoid always allocating the `Value` in cases where we have invisible values. This still removes the extra `collector.get` via `computeIfAbsent`